pkg/validation/validation-rule.md

```diff
@@ -14,7 +14,8 @@ The validation will be done as part of schema validation, the rule will be intro
 
 ### Endpoints:
 - all the endpoint names are unique across components
-- all the endpoint ports are unique across components. Only exception: container component with `dedicatedpod=true`
+- endpoint ports must be unique across components -- two components cannot have the same target port, but one component may have two endpoints with the same target port. This restriction does not apply to container components with `dedicatedPod` set to `true`.
```

Comment:
If we allow the use of the same port on one container, why not allow the same ports on two different containers? Isn't basically the same situation as in most cases the containers will be in on Pod?


We don't need to block this PR on this.
We can revisit it in a separate issue.

Reply:
Kubernetes won't allow multiple containerPorts within the same pod, so allowing the same port on different containers will result in the pod being rejected as invalid. For endpoints on the same container, only one containerPort is created (and only one service within the pod can listen on that port).

The endpoint defines more information than just the port. In particular, we use endpoints to define URL paths, protocols, and whether a route should be created. This results in multiple ingresses/routes that map to 1 or more services, that end up all pointing at one port in the container. An example would be something like
```
https://url1.com/path1 ───────────┐
                                  ▼
                           service port 8080 ──► container port 8080
                                  ▲
https://url2.com/path2 ───────────┘
```
Perhaps the ultimately better solution (requiring a new APIVersion) would be to extend endpoints to allow configuring something like this in one endpoint entry.

Reply:
> Kubernetes won't allow multiple containerPorts within the same pod, so allowing the same port on different containers will result in the pod being rejected as invalid. For endpoints on the same container, only one containerPort is created (and only one service within the pod can listen on that port).

oh, right.
In this case, we have quite a broken implementation of endpoints in devfile/library :-/
Currently, it maps every endpoint into the container port.
By re-introducing this change devfile/library will have to start doing smarter things with endpoints than just blindly mapping 1:1 to container ports.
/cc @elsony @yangcao77 

I think that this is another proof of how badly we need a common layer that controls the basic principles of how Devfile is translated to k8s resources. 
If we can't have it on code level in for of some shared library then at least we need to have documentation describing the ground rules of how devfile is translated to k8s resources. Otherwise, problems like this will keep coming.

Reply:
Yes. I have aware that library need to make change accordingly. But library can still use the endpoint name, if multiple endpoints uses same port, library can use the first endpoint uses that port. Or Odo want to change back to the old behavior? all use `port-<portNumber>` instead?